src/main/java/com/commercetools/pspadapter/paymentHandler/impl/PaymentHandler.java

```diff
@@ -177,36 +231,69 @@ public PaymentHandleResponse executePayment(@Nonnull String paypalPlusPaymentId,
-                SetInterfaceId.of(newPpPayment.getId()));
-        return ctpFacade.getPaymentService().updatePayment(ctpPaymentId, updateActions)
-                .thenApply(ignore -> newPpPayment);
+    protected List<UpdateAction<io.sphere.sdk.payments.Payment>> getApprovalUrlAndInterfaceIdAction(Payment paypalPayment) {
+        return asList(
+                SetCustomField.ofObject(APPROVAL_URL, PaymentMapper.getApprovalUrl(paypalPayment).orElse("")),
+                SetInterfaceId.of(paypalPayment.getId()));
+    }
+
+    /**
+     * Private methods
+     **/
+
+    private AddInterfaceInteraction createAddInterfaceInteractionAction(PayPalModel model,
+                                                                        InterfaceInteractionType type) {
+        Gson gson = new GsonBuilder()
+                .setFieldNamingPolicy(FieldNamingPolicy.IDENTITY)
+                .disableHtmlEscaping()
+                .create();
+        String json = gson.toJson(model);
+        return AddInterfaceInteraction.ofTypeKeyAndObjects(type.getInterfaceKey(),
+                ImmutableMap.of(type.getValueFieldName(), json,
+                        "timestamp", ZonedDateTime.now()));
     }
 
     private CompletionStage<PaymentHandleResponse> executePaymentAndCreateTxn(@Nonnull String paypalPlusPaymentId,
                                                                               @Nonnull String paypalPlusPayerId,
                                                                               @Nonnull io.sphere.sdk.payments.Payment ctpPayment) {
-        return paypalPlusFacade.getPaymentService().execute(new Payment().setId(paypalPlusPaymentId),
-                new PaymentExecution().setPayerId(paypalPlusPayerId))
-                .thenCompose(paypalPayment -> {
-                    if (PaypalPlusPaymentStates.APPROVED.equals(paypalPayment.getState())) {
-                        return createChargeTransaction(paypalPayment, ctpPayment.getId(), TransactionState.SUCCESS);
-                    } else if (PaypalPlusPaymentStates.CREATED.equals(paypalPayment.getState())) {
-                        return createChargeTransaction(paypalPayment, ctpPayment.getId(), TransactionState.PENDING);
-                    } else {
-                        throw new PaypalPlusException(format("Error when approving payment [%s], current state=[%s]",
-                                paypalPlusPaymentId, paypalPayment.getState()));
-                    }
+        PaymentExecution paymentExecution = new PaymentExecution().setPayerId(paypalPlusPayerId);
+
+        AddInterfaceInteraction interactionAction = createAddInterfaceInteractionAction(paymentExecution, REQUEST);
+        return ctpFacade.getPaymentService()
+                // 1. add execute request to the ctp payment as interaction interface
+                .updatePayment(ctpPayment.getId(), Collections.singletonList(interactionAction))
+                .thenCompose(payment -> {
```

Comment:
maybe instead of all the inline comments split the futures into own methods when method naming self explaining what is happening instead of comments.

Reply:
They're already in their own methods, but I just want to have this 1. - 4. steps, so you know that it goes like this.

Reply:
+1 on Hesham's point

Reply:
doesn't seem that they are on their own methods with names. For example:


` return ctpFacade.getPaymentService().updatePayment(payment, Collections.singletonList(addInteractionAction))` doesn't explain to me what is stated in it's comment: `// 3. add paypal response to the ctp payment as interaction interface`

Reply:
@heshamMassoud @lojzatran see my comment https://github.com/commercetools/commercetools-paypal-plus-integration/pull/49#discussion_r136528547

Reply:
https://github.com/commercetools/commercetools-paypal-plus-integration/pull/49/commits/8b8630fd3fa0bedf3778d831871058aff3734c62